AlphaWallet/Core/CoinTicker/CoinTickersResolver.swift

```diff
@@ -0,0 +1,243 @@
+                    loadChartHistory(period: period, ticker: ticker)
+                }.then(on: CoinTickersResolver.queue, { result -> Promise<ChartHistory> in
+                    guard !result.prices.isEmpty else { return .value(result) }
+
+                    var valueToSet = self.history[ticker] ?? [:]
+                    valueToSet[period] = result
+
+                    self.history[ticker] = valueToSet
+                    return .value(result)
+                })
+            }
+        } else {
+            return .value(.empty)
+        }
+    }
+
+    private func coinTickersPromise(enabledObject: [TokenMappedToTicker]) -> Promise<[AddressToPRCServerKey: CoinTicker]> {
+        guard fetchTickersIsAvailable else {
+            return .value(tickers)
+        }
+
+        guard !isFetchingPrices else {
+            return .init(error: AnyError.alreadyFetchingPrices)
+        }
+
+        let config = self.config
+
+        isFetchingPrices = true
+
+        return firstly {
+            fetchCoinTikerIds()
+        }.compactMap(on: CoinTickersResolver.queue, { ids -> [MappedCoinTickerId] in
+            let mappedTokensToCoinTickerIds = enabledObject.compactMap { tokenObject -> MappedCoinTickerId? in
+                if let id = ids.first(where: { $0.isAvailable(for: tokenObject) }) {
+                    return MappedCoinTickerId(id: id.id, contractAddress: tokenObject.contractAddress, server: tokenObject.server)
+                } else {
+                    return nil
+                }
+            }
+            return mappedTokensToCoinTickerIds
+        }).then(on: CoinTickersResolver.queue, { mapped -> Promise<[AddressToPRCServerKey: CoinTicker]> in
+            let ids: String = Set(mapped).map { $0.coinTickerId }.joined(separator: ",")
+            
+            return firstly {
+                self.provider.request(.priceTokens(config: config, ids: ids, currency: Constants.Currency.usd))
+            }.map(on: CoinTickersResolver.queue, { response -> [AddressToPRCServerKey: CoinTicker] in
+                let tickers = try response.map([CoinTicker].self, using: JSONDecoder())
+
+                var resultTickers: [AddressToPRCServerKey: CoinTicker] = [:]
+                for ticker in tickers {
+                    if let value = mapped.first(where: { $0.coinTickerId == ticker.id }) {
+                        let key = AddressToPRCServerKey(address: value.contractAddress, server: value.server)
+                        resultTickers[key] = ticker
+                    }
+                }
+                return resultTickers
+            }).recover(on: CoinTickersResolver.queue, { _ -> Promise<[AddressToPRCServerKey: CoinTicker]> in
+                return .value(self.tickers)
+            })
+        }).ensure(on: CoinTickersResolver.queue, { [weak self] in
+            self?.isFetchingPrices = false
+        })
+    }
+}
+
+private extension CoinTickersResolver {
+    struct CoinTickerId: Codable {
```

Comment:
How about just `Ticker`? The `Id` doesn't communicate its purpose well, `Ticker` isn't perfect, but it's a private type, so..